numpy/core/src/multiarray/number.c

```diff
@@ -292,10 +299,109 @@ PyArray_GenericInplaceUnaryFunction(PyArrayObject *m1, PyObject *op)
+     * element-wise basis for the concatented unicode elements;
+     * perhaps it is possible to iterate through and determine
+     * the largest result element and then pre-allocate to
+     * U(largest) type; not quite sure how to do that so I
+     * instead use a more flexible object array and then force
+     * cast back to NPY_UNICODE before returning
+     */
+
+    ret = PyArray_NewLikeArray((PyArrayObject *)self,
+                               NPY_ANYORDER,
+                               PyArray_DescrFromType(NPY_OBJECT),
+                               0);
+
+    self_iter = (PyArrayIterObject *)PyArray_IterNew((PyObject *)self);
+    other_iter = (PyArrayIterObject *)PyArray_IterNew((PyObject *)other);
+    ret_iter = (PyArrayIterObject *)PyArray_IterNew((PyObject *)ret);
+
+    while (other_iter->index < other_iter->size) {
+        /* retrieve item from self */
+        item = PyArray_GETITEM((PyArrayObject *)self,
+                               PyArray_ITER_DATA(self_iter));
+        /* retrieve item from other array in concat op */
+        other_item = PyArray_GETITEM((PyArrayObject *)other,
+                                     PyArray_ITER_DATA(other_iter));
+
+        /*
+         * concat the two Unicode PyObject elements and
+         * place them at appropriate index in ret
+         */
+        combined = PyUnicode_Concat(item, other_item);
+        PyArray_SETITEM((PyArrayObject *)ret,
+                         PyArray_ITER_DATA(ret_iter),
+                         combined);
+        /* probably need XDECREFs */
+        PyArray_ITER_NEXT(self_iter);
+        PyArray_ITER_NEXT(other_iter);
+        PyArray_ITER_NEXT(ret_iter);
+    }
+
+    Py_INCREF(ret);
+
+    /* perhaps a little ugly to have to force cast back */
+    ret2 = PyArray_FROM_OTF(ret,
+                            NPY_UNICODE,
+                            NPY_ARRAY_FORCECAST);
+    return ret2;
+}
+
 static PyObject *
 array_add(PyArrayObject *m1, PyObject *m2)
 {
     PyObject *res;
+    /*
+     * this function is pointed to by the nb_add
+     * slot used by np.ndarray, and apparently also
+     * for Unicode arrays
+     * I suppose long-term we want to have a graceful structure
+     * for specifying these slots instead of intercepting
+     * inside the general __add__ function pointer as I've done here
+     */
+    if (PyArray_Check(m1)) {
+        /*
+         * specialized dispatch for nb_add for different array
+         * types is now guarded above against the case where
+         * PyArrayObject m1 is not a PyArrayObject, which seems
+         * like a sensible requirement for __add__ for np.ndarray
+         * type struct, but so far we haven't been checking this
+         */
+        if (PyArray_DESCR(m1)->type_num == NPY_UNICODE &&
+            PyArray_DESCR((PyArrayObject *)m2)->type_num == NPY_UNICODE) {
+            return unicodetype_concat((PyObject *)m1, m2);
+        }
+    }
```

Comment:
What happens if `PyArray_Check(m2)` is false? For instance, try adding another statement in the new test: `test + u'abc'`